File: src/java/3_StructuralPatterns/3_Composite/CompositeExercise.java
package com.activemesa.structural.composite.exercise;

import java.util.*;

interface ValueContainer extends Iterable<Integer> {}

class SingleValue implements ValueContainer
{
  public int value;

  public SingleValue(int value)
  {
    this.value = value;
  }

  @Override
  public Iterator<Integer> iterator()
  {
    return Collections.singleton(value).iterator();
  }
}

class ManyValues extends ArrayList<Integer> implements ValueContainer
{
}


class MyList extends ArrayList<ValueContainer>
{
  public MyList(Collection<? extends ValueContainer> c)
  {
    super(c);
  }

  public int sum()
  {
    int result = 0;
    for (ValueContainer c : this)
    {
      for (int i : c)
      {
        result += i;
      }
    }
    return result;
  }
}

File: src/js/3_StructuralPatterns/3_Composite/composite_exercise.js
class SingleValue
{
  constructor(value)
  {
    this.value = value;
  }

  [Symbol.iterator]()
  {
    let returned = false;
    return {
      next: () => ({
        value: this.value,
        done: returned++
      })
    }
  }
}

class ManyValues extends Array {}

let sum = function(containers)
{
  let result = 0;
  for (let c of containers)
    for (let i of c)
      result += i;
  return result;
};

describe('composite', function()
{
  it('should sum up different objects', function()
  {
    let singleValue = new SingleValue(11);
    let otherValues = new ManyValues();
    otherValues.push(22);
    otherValues.push(33);
    expect(sum([singleValue, otherValues])).toEqual(66);
  });
});